appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt

```diff
@@ -0,0 +1,311 @@
+    private val experienceRenderer: ExperienceRenderer = mockk {
+        coEvery { getStateFlow(RenderContext.Modal) } returns experienceStates
+    }
+
+    private val actionProcessor: ActionProcessor = mockk(relaxed = true)
+
+    private val onDismiss: () -> Unit = mockk(relaxed = true)
+
+    private lateinit var viewModel: AppcuesViewModel
+
+    private val uiStates = arrayListOf<UIState>()
+
+    private lateinit var stateJob: Job
+
+    @Before
+    fun setup() {
+        Dispatchers.setMain(UnconfinedTestDispatcher())
+
+        viewModel = AppcuesViewModel(renderContext, coroutineScope, experienceRenderer, actionProcessor, onDismiss).apply {
+            stateJob = viewModelScope.launch { uiState.collect { uiStates.add(it) } }
+        }
+
+        // clear getStateFlow from history
+        excludeRecords { experienceRenderer.getStateFlow(renderContext) }
+    }
+
+    @After
+    fun tearDown() {
+        stateJob.cancel()
+
+        Dispatchers.resetMain()
+    }
+
+    @Test
+    fun `first uiState SHOULD be Idle`() = runTest {
+        assertThat(uiStates).hasSize(1)
+        assertThat(uiStates[0]).isInstanceOf(Idle::class.java)
+    }
+
+    @Test
+    fun `BeginningStep SHOULD emit Rendering`() = runTest {
+        // GIVEN
+        val mockStepContainer: StepContainer = mockk()
+        val mockExperience: Experience = mockk(relaxed = true) {
+            every { groupLookup } returns hashMapOf(0 to 0)
+            every { stepIndexLookup } returns hashMapOf(0 to 1234)
+            every { stepContainers } returns listOf(mockStepContainer)
+        }
+        val isFirst = true
+        val mockMetadata: Map<String, Any?> = hashMapOf()
+        // WHEN
+        experienceStates.emit(BeginningStep(mockExperience, 0, isFirst, mockMetadata))
+        // THEN
+        assertThat(uiStates).hasSize(2)
+        with(uiStates[1] as Rendering) {
+            assertThat(experience).isEqualTo(mockExperience)
+            assertThat(stepContainer).isEqualTo(mockStepContainer)
+            assertThat(position).isEqualTo(1234)
+            assertThat(flatStepIndex).isEqualTo(0)
+            assertThat(metadata).isEqualTo(mockMetadata)
+        }
+    }
+
+    @Test
+    fun `BeginningStep SHOULD NOT emit Rendering WHEN flatStepIndex get NULL on groupLookup`() = runTest {
+        // GIVEN
+        val mockStepContainer: StepContainer = mockk()
+        val mockExperience: Experience = mockk(relaxed = true) {
+            // different index
+            every { groupLookup } returns hashMapOf(3 to 0)
+            every { stepIndexLookup } returns hashMapOf(0 to 1234)
+            every { stepContainers } returns listOf(mockStepContainer)
+        }
+        val isFirst = true
+        val mockMetadata: Map<String, Any?> = hashMapOf()
+        // WHEN (flatStepIndex is 2 here)
+        experienceStates.emit(BeginningStep(mockExperience, 0, isFirst, mockMetadata))
+        // THEN
+        assertThat(uiStates).hasSize(1)
```

Comment:
for my own understanding, where does uiStates get reset between each test?

Reply:
yes, each run the list is a new one